csharp/lib/src/lib.rs

```diff
@@ -26,37 +28,111 @@ pub struct Client {
     runtime: Runtime,
 }
 
-fn create_connection_request(
-    host: String,
-    port: u32,
-    use_tls: bool,
+/// Convert raw C string to a rust string.
+///
+/// # Safety
+///
+/// * `ptr` must be able to be safely casted to a valid `CString` via `CString::from_raw` if `ptr` is not null. See the safety documentation of [`std::ffi::CString::from_raw`](https://doc.rust-lang.org/std/ffi/struct.CString.html#method.from_raw).
+unsafe fn ptr_to_str(ptr: *const c_char) -> String {
+    if ptr as i64 != 0 {
+        unsafe { CStr::from_ptr(ptr) }.to_str().unwrap().into()
+    } else {
+        "".into()
+    }
+}
+
+/// Convert raw array pointer to a vector of [`NodeAddress`](NodeAddress)es.
+///
+/// # Safety
+///
+/// * `len` must not be greater than `isize::MAX`. See the safety documentation of [`std::slice::from_raw_parts`](https://doc.rust-lang.org/std/slice/fn.from_raw_parts.html).
+/// * `data` must not be null.
+/// * `data` must point to `len` consecutive properly initialized [`NodeAddress`](NodeAddress) structs.
+/// * Each [`NodeAddress`](NodeAddress) dereferenced by `data` must contain a valid string pointer. See the safety documentation of [`ptr_to_str`](ptr_to_str).
+#[allow(rustdoc::redundant_explicit_links)]
+unsafe fn node_addresses_to_proto(
+    data: *const *const NodeAddress,
+    len: usize,
+) -> Vec<connection_request::NodeAddress> {
+    unsafe { std::slice::from_raw_parts(data as *mut NodeAddress, len) }
+        .iter()
+        .map(|addr| {
+            let mut address_info = connection_request::NodeAddress::new();
+            address_info.host = unsafe { ptr_to_str(addr.host) }.into();
+            address_info.port = addr.port as u32;
+            address_info
+        })
+        .collect()
+}
+
+/// Convert connection configuration to a corresponding protobuf object.
+///
+/// # Safety
+///
+/// * `config` must not be null.
+/// * `config` must be a valid pointer to a [`ConnectionConfig`](ConnectionConfig) struct. See the safety documentation of [`std::ptr`](https://doc.rust-lang.org/std/ptr/index.html#safety).
```

Comment:
While it is better safe than sorry, I think that we don't need to tell people to look at the safety docs for raw pointers. This might be a little overkill.